### ISSUE TITLE
Scale line shows incorrect sizes.

### ISSUE THREAD

**Issue by jonathanAK** [openlayers/openlayers#10529]
The scale line control (ol/control/ScaleLine), assumes a pixel size of 0.28mm.
This creates very large errors in the scale.
On mobile devices we tested we found the average error being more that 50%.

Passing the screen density as an option when creating the control should give an easy solution.

In addition the function converts mm to inch (X25.4) and then inch to meters(X39.37), since these are only used internally it seems redundant to introduce these errors instead of converting mm to meters directly.

  getScaleForResolution() {
    const resolution = this.getMap().getView().getResolution();
    const dpi = **25.4** / 0.28;
    const mpu = this.viewState_.projection.getMetersPerUnit();
    const inchesPerMeter = **39.37**;
    return parseFloat(resolution.toString()) * mpu * inchesPerMeter * dpi;
  }

**Comment by mike-000** [comment]
There is no way for a browser to determine the precise pixel size of a device. However if exporting to a PDF with specified dpi and with the map resolution set for a specified scale at that dpi there should be a way to reflect that in the scalebar https://codesandbox.io/s/export-pdf-r2m3x  Developers could also use that to better reflect that characteristics of target devices.

**Comment by ldv-dev** [comment]
This has been discussed here before, and still bugs me a bit.

I think it all comes to de dpi values of ( map / resolutions / browser / canvas / etc)

- the ol scale line control uses a fixed dpi of 25.4 / 0.28 = 90,714.
- if the map is created with pixelRatio: 1 it should not change on different devices and should make math easier. 
- if we export the canvas/dom/(map), it exports an image from the browser canvas that has a fixed 96dpi, so map/scales are created with 90,714 but the export comes on 96, so you need to adjust the export image size from 90,714 to 96, otherwise you get an image with the scale calculated to 90.714 but it actually is at 96.

If you want fixed scale values for each zoom (1:1000, 1:2000, 1:5000, ...) so you can export a map image at 1:1000, you need to match each zoom level each scale value, using a dpi value of 25.4 / 0.28, like this each zoom will be even with the scale line control.


All this because "DPI defined by OGC specification (dpi assumes 0.28mm as the physical distance of a pixel)." witch since the browser is at 96dpi, and since OL is to show layers/data on browsers, shouldn't the zoom lvls be visually more accurate using 96dpi?

I would like to build a map/resolutions at 96dpi, with a scaleLineControl at 96dpi, the exported map to png/pdf will be exported at 96(browser), and if the pdf is created with 96dpi, 1cm on the screen should correspond to a even scale line value and after the export with should be 1cm on the pdf as well, if printed with 96dpi also, we should be getting the "same" approximated 1cm on paper.

All this is a bit confusing and i am stretching quite a bit here...

**Comment by mike-000** [comment]
If you are exporting to a 300 dpi printer (or a PDF intended for one such as in the example https://openlayers.org/en/latest/examples/export-pdf.html) the ScaleLine's dpi should be 300 as the export has been correspondingly enlarged.  Also the resolution used in the ScaleLine is currently incorrect unless you are at the equator - it needs to be adjusted for the point resolution of the view center.  Here is my previous example patched to show the correct scale when exported  https://codesandbox.io/s/export-pdf-wquew  (the controls are not yet correctly positioned on the export - that is a limitation of the basic example which wasn't intended to include controls).

**Comment by ldv-dev** [comment]
ok... 

1) so you adjusted/replaced this.getMap().getView().getResolution() with getPointResolution() using the center point of the map in order to have a more accurate resolution value, right?
also override the consts on ScaleLine so it would auto adjust to the chosen DPI value?

2) By default ScaleLine uses const dpi = 25.4 / 0.28, do you know if the map when created and calculating the resolution for all the zoom levels also uses a dpi = 25.4 / 0.28?

if so? if we want to try to approximate what we see in our screen to reallity/paper can or should we calculate map resolutions using dpi = 25.4 / 0.28 OR dpi = 96(browser)?

thanks in advance

**Comment by DaniEll-AT** [comment]
**TLDR**: Don't calculate the factor between real world distance and screen distance, because it's currently not accurately possible. Instead show a bar with a label indicating the correspondending real world distance.

**Reasoning**
As i see it, all boils down to
- current browsers don't tell websites the screens dpi
- we can only guess about the screens dpi to show a scale-number like 1:1000
- showing a scale like this is only accurately possible, if someone tells us the exact dpi to expect (like in "print with 300dpi")

What we can do
- resolution tells us, that 1 px is [x] meters
- we can calculate, how much px [y] meters are
- we could show a bar [z] px wide, with a label [y] meters

That would leave just one open questions
- how wide should that bar be (min, max px)?

Just my 2¢

**Comment by ldv-dev** [comment]
> **TLDR**: Don't calculate the factor between real world distance and screen distance, because it's currently not accurately possible. Instead show a bar with a label indicating the matching real world distance.
> 
> **Reasoning**
> As i see it, all boils down to
> 
> * current browsers don't tell websites the screens dpi
> * we can only guess about the screens dpi to show a scale-number like 1:1000
> * showing a scale like this is only accurately possible, if someone tells us the exact dpi to expect (like in "print with 300dpi")
> 
> What we can do
> 
> * resolution tells us, that 1 px is [x] meters
> * we can calculate, how much px [y] meters are
> * we could show a bar [z] px wide, with a label [y] meters
> 
> That would leave just one open questions
> 
> * how wide should that bar be (min, max px)?
> 
> Just my 2¢

I don't expect to have a 1:1 match, screen to real distance, but if it could be adjusted to be as accurate as possible it would be great.

About "current browsers don't tell websites the screens dpi", yes but as long as create the map with pixelRatio: 1, my browser return fixed 96 dpi or i am wrong?

**Comment by DaniEll-AT** [comment]
Your browser renders pixels to the graphic cards frame buffer.
The dpi and real world size of the image is a consequence of screen resolution and monitor dimensions (22" screen, 1680 x 1050 px => x dpi, what about beamers :innocent: ?)

**Comment by mike-000** [comment]
The constructor already has a `minWidth` option, and the actual length is determined by the next distance on the ground which exceeds the `minWidth` (the steps being 1, 2, 5, 10, 20, 50, 100 km, etc) and I don't see any need to change that.

I agree with @DaniEll-AT - while the browser may return a nominal 96 dpi the actual dpi would depend on the size of the screen and cannot distinguish between a 22 inch 1080p monitor a 32 inch TV screen, so showing a scale usually isn't very useful.

However for printing (and perhaps for some mobile devices with known capabilities) a scale is useful there should be some way of overriding the default (a `dpi` option in the constructor and/or a `setDpi()` method).  The resolution **must** also be corrected for point resolution (as it already is elsewhere in the control - if the length of the bar changes as you pan the map so should the scale!).  An additional example showing how to print to a specified scale (with scalebar suitably positioned) would be useful (a source which includes a 1 or 10 km grid such as Ordnance Survey or SwissTopo would provide a good demonstration).

**Comment by ldv-dev** [comment]
After further inspection on measuring Lengths/Areas the ScaleLine Control uses the Unit from the map projection (my case EPSG:3857, metric), OK.

But when the size of each bar is calculated (based on resolution), the resolution value is recalculated based on "getPointResolution() from 'proj.js' witch converts the center point to EPSG:4326.

Doesn't this makes a "Ellipsoidal" based scaleLine on a Cartesian Map?

Or all distances should be calculated on Ellipsoidal bases like "var length = ol.sphere.getLength(lineGeom)"?

**Comment by mike-000** [comment]
The length of the scale line adjusted for point resolution is the same as the radius of the "ellipoidal" (OpenLayers uses spherical math) scale at the map center along the x axis https://codesandbox.io/s/scalerings-4zfij
